misc/man/rrr_sub.1
.Dd 2026-01-02
.TH RRR_SUB 1
.SH NAME
Read Route Record is a dynamic module-based message handling program
for collecting, buffering, processing and storing messages and measurements.
.PP
.B rrr_sub
subscribes to topics on an MQTT broker and prints out messages received.
Both messages containing RRR messages and plain data are supported.
.SH SYNOPSIS
.B rrr_sub [TOPIC]...
.Dl [-B|--broker[=]BROKER] (defaults to localhost)
.Dl [-P|--port[=]PORT] (defaults to 1883)
.Dl [-n|--nice-json]
.Dl [-Q|--qos[=]QoS] (defaults to 0)
.Dl [-5|--protocol-v5]
.Dl [-3|--protocol-v3]
.Dl [-j|--json]
.Dl [-l|--loglevel-translation]
.Dl [-b|--banner]
.Dl [-e|--environment-file[=]ENVIRONMENT FILE]
.Dl [-d|--debuglevel[=]DEBUG FLAGS]
.Dl [-D|--debuglevel-on-exit[=]DEBUG FLAGS]
.Dl [-h|--help]
.Dl [-v|--version]

.SH DESCRIPTION
This program reads plain or RRR messages from an MQTT broker.

The type of message is auto-detected. If it appers to be an RRR message, checksums will be validated.

If the connection to the MQTT broker is lost or cannot be established,
.B rrr_sub
will attempt to reconnect indefinately.

.SH OPTIONS
.IP [TOPIC]...
One or more topic filters to subscribe to may be specified. If none are given, no subriptions are made.
.IP -B|--broker[=]BROKER
Set the address of the broker to use. 'localhost' is used if left unset.
.IP -P|--port[=]PORT
Set the port number to use when connecting to the broker. Port 1883 is used if left unset.
.IP -n|--nice-json
Attempt to interpret received messages (which plain MQTT data messages) as JSON and produce a formatted output.
.IP -Q|--qos[=]QoS
Set the Quality of Service of subscriptions to 2, 1 or 0. The actual QoS the broker accepts for each subscription might be different.
.IP -5|--protocol-v5
.IP -3|--protocol-v3
Enable protocol version 3.1.1 and/or 5. The default behaviour is that 5 is tried first, then 3.1.1 if the connection fails.
.IP -j|--json
If given, logging will be printed in JSON format. Refer to
.Xr rrr_logd(1)
for details.
.IP -l|--loglevel-translation
Translate output loglevels to RFC5424 format. Internal loglevels will be translated so that error messages become ERROR
and other messages DEBUG (loglevels 3 and 7).
.IP -b|--banner
Print RRR banner before starting.
.IP -e|--environment-file
Specify an environment file to read in before the program starts. See
.Xr rrr(1)
for details.
.IP -d|--debuglevel[=]FLAGS
.IP -D|--debuglevel_on_exit[=]FLAGS
Enable debug messages.

The debuglevel is produced by adding one or more of the following numbers together:
.Bl -tag -width -indent
.It 0
Severe errors (always printed)
.It 1
General information about communication with the MQTT broker.
.It 2
Dumps any array data from an RRR array message and status of subscriptions.
.It 4
Detailed information about communication with the MQTT broker.
.It 32
Dumps any data from an RRR data message.
.It 64
Prints debug information about sockets and file reading.
.El

Other debuglevels specified in
.Xr rrr(1)
can be used but might not be useful.

The parameter
.B debuglevel_on_exit
has not effect.

.SH RETURN VALUE
.B rrr_sub
returns 0 on success or 1 on errors.

.SH DEBUG FLAGS
Refer to
.Xr rrr(1)
for possible debug flags.

.SH SEE ALSO
.Xr rrr(1)


debian/changelog
rrr (1.40-2) trixie; urgency=medium

  * RUNNING DEVELOPMENT RELEASE
  * Changes:
    * Require build with JSON-C bindings

  * Features:
    * The rrr_sub MQTT utility supports nice print of JSON messages
    * New module for GPIO control

 -- Atle Solbakken <atle@goliathdns.no>  Thu, 18 Dec 2025 09:30:54 +0100

rrr (1.39-1) trixie; urgency=medium

  * Changes:
    * In httpclient, the option http_low_priority_put no longer affects
      messages from message DB. http_msgdb_low_priority_put replaces
      this case.

  * Features:
    * In httpclient, add ability to prioritize failed PUT request when they are
      retried from message DB storage
    * ArtNet module in main release thread, disabled by default

  * Improvements:
    * Better logging of problems with forks crashing

  * Misc:
    * New cmodule with GTK example code

 -- Atle Solbakken <atle@goliathdns.no>  Thu, 18 Dec 2025 08:39:26 +0100

rrr (1.37-1) trixie; urgency=medium

  * Changes:
    * New version will be built against Debian Trixie only except in case of
      critical bugfixes

  * Features:
    * New methods on JS Message object: data_as_object and data_as_utf8
    * Support for Node 23
    * Support for Debian Trixie
    * Log daemon may start a wrapper process for further processing of
      messages prior to them being printed.
    * Configurable sub instances in cmodules
    * Native JSON logging from log daemon
    * Ability to send single ModBus commands (oneshot commands)

  * Improvements:
    * In httpserver, use blob array types when receiving
      application/octet-stream type

  * Fixes:
    * BAIL in method and route definitions not working correctly, caused
      program exit when triggered
    * FALSE value in method and route definitions not working, TRUE was pushed
      instead
    * Missing check for overshoot and undershoot while fading in ArtNet
    * Incorrect counting of FIFO buffer entries in certain situations, could
      possibly result in buffers over time being reported as full when they
      were not. The problem only affected counting of retained messages in MQTT
      broker.
    * Misc. compilation problems using libc++ on FreeBSD, which is required
      then V8 is used on this platform
    * MQTT client not always reconnecting as configured but restarted
    * JS incorrectly loading the same module multiple times if included
      from different scripts

 -- Atle Solbakken <atle@goliathdns.no>  Fri, 29 Aug 2025 14:02:39 +0200

rrr (1.35-1) bookworm; urgency=medium

  * Features:
    * Configuration option in http client to trap any header field

  * Fixes:
    * Crashes in MQTT for malformed packets with tailing payload
    * Bad MQTT CONNECT packet assembling for V5 when will flag was not set
    * Log messages to stats engine  getting lost due to events delivered to
      incorrect event queue
    * Memory leak in Python3 where values added to array had incorrect
      reference count causing memory leaks
    * Broken parsing of leading spaces in istr and ustr types in arrays
    * Crash in array tree import for division overflow in conditions,
      emit soft error instead.

 -- Atle Solbakken <atle@goliathdns.no>  Wed, 27 Nov 2024 18:32:12 +0100

rrr (1.33-1) bookworm; urgency=medium

  * Features:
    * Log socket daemon with ability to collect logs from multiple RRR
      processes

  * Improvements:
    * Remove all references to http2 when RRR is built without nghttp2, in
      particular avoid noisy notification caused by inconsistencies in the
      code
    * Speed up tests by running them in parallel
    * In httpclient, by default ignore all meta fields from httpserver which
      should usually not be added as request fields.
    * In httpclient, prevent using // endpoint when using endpoint from topic
      and url is /

  * Fixes:
    * Remove stale python directory from source causing build to fail in some
      cases
    * Misc. hangs and crashes when parsing invalid configuration files
    * Misc. crashes in HTTP handling

 -- Atle Solbakken <atle@goliathdns.no>  Fri, 01 Nov 2024 07:02:09 +0100

rrr (1.31-5) bookworm; urgency=medium

  * Fixes:
    * Preflight headers not sendt for OPTIONS request
    * Move setproctitle() call to ensure that it cannot clear environment
      variables prior to them being accessed during initialization.

 -- Atle Solbakken <atle@goliathdns.no>  Mon, 12 Aug 2024 13:05:28 +0200

rrr (1.31-1) bookworm; urgency=medium

  * Features:
    * New option http_server_decode_endpoint to decode the endpoint field
      when receiving full requests.
    * Support HTTP3 over QUIC in HTTP server and client.
    * Start RRR in single thread mode which is useful for debugging and
      on slower hardware.
    * Support for more ModBus functions
      - By Fredrik Rosenqvist <rosenqvistfredrik@hotmail.com>
    * New method rrr_array.has() in the Python module to check if an array tag
      exists.
      - By Fredrik Rosenqvist <rosenqvistfredrik@hotmail.com>
    * New option http_server_topic_format to include more information about a
      request in the topic of messages generated by HTTP server module.
      - By Fredrik Rosenqvist <rosenqvistfredrik@hotmail.com>

  * Improvements:
    * Implement polite close of connections during shutdown for HTTP1
      and HTTP2 in HTTP server.
    * Optimization in message database to avoid excessive chdir calls.

  * Fixes:
    * Broken TLS mode in MQTT client.
    * All settings used check not working correctly for JS module.

 -- Atle Solbakken <atle@goliathdns.no>  Wed, 29 May 2024 08:43:01 +0200

rrr (1.29-2) bookworm; urgency=medium

  * Changes:
    * In route definition, the I operator is replaced with D. Existing programs
      must be updated to prevent warning from being printed.
    * Changed signatures for processor functions by adding a new last argument
      'method'. Existing scripts MAY BREAK and should be updated. Refer to

src/rrr_sub.c
/*

Read Route Record

Copyright (C) 2022-2026 Atle Solbakken atle@goliathdns.no

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/

#include <stdlib.h>
#include <stdio.h>

#include "../build_timestamp.h"
#include "lib/rrr_config.h"
#include "lib/util/macro_utils.h"
#include "main.h"
#include "lib/allocator.h"
#include "lib/common.h"
#include "lib/version.h"
#include "lib/log.h"
#include "lib/rrr_strerror.h"
#include "lib/cmdlineparser/cmdline.h"
#include "lib/mqtt/mqtt_client.h"
#include "lib/mqtt/mqtt_subscription.h"
#include "lib/mqtt/mqtt_packet.h"
#include "lib/mqtt/mqtt_payload.h"
#include "lib/socket/rrr_socket.h"
#include "lib/net_transport/net_transport_config.h"
#include "lib/util/rrr_time.h"
#include "lib/util/arguments.h"
#include "lib/messages/msg.h"
#include "lib/messages/msg_msg.h"
#include "lib/messages/msg_dump.h"
#include "lib/json/json.h"

#define RRR_MQTT_SUB_TOPICS_MAX 64
#define RRR_MQTT_DISCONNECT_TIMEOUT_S 5

RRR_CONFIG_DEFINE_DEFAULT_LOG_PREFIX("rrr_sub");

static const struct cmd_arg_rule cmd_rules[] = {
        {CMD_ARG_FLAG_NO_FLAG_MULTI,   '\0',   "topic",                "[TOPIC]..."},
        {CMD_ARG_FLAG_HAS_ARGUMENT,    'B',    "broker",               "[-B|--broker]"},
        {CMD_ARG_FLAG_HAS_ARGUMENT,    'P',    "port",                 "[-P|--port]"},
	{CMD_ARG_FLAG_HAS_ARGUMENT,    'Q',    "qos",                  "[-Q|--qos]"},
	{0,                            'n',    "nice-json",            "[-n|--nice-json]"},
	{0,                            '3',    "protocol-v3",          "[-3|--protocol-v3]"},
	{0,                            '5',    "protocol-v5",          "[-5|--protocol-v5]"},
	{0,                            'j',    "json",                 "[-j|--json]"},
        {0,                            'l',    "loglevel-translation", "[-l|--loglevel-translation]"},
        {0,                            'b',    "banner",               "[-b|--banner]"},
        {CMD_ARG_FLAG_HAS_ARGUMENT,    'e',    "environment-file",     "[-e|--environment-file[=]ENVIRONMENT FILE]"},
        {CMD_ARG_FLAG_HAS_ARGUMENT,    'd',    "debuglevel",           "[-d|--debuglevel[=]DEBUG FLAGS]"},
        {CMD_ARG_FLAG_HAS_ARGUMENT,    'D',    "debuglevel-on-exit",   "[-D|--debuglevel-on-exit[=]DEBUG FLAGS]"},
        {0,                            'h',    "help",                 "[-h|--help]"},
        {0,                            'v',    "version",              "[-v|--version]"},
        {0,                            '\0',    NULL,                  NULL}
};

struct rrr_sub_data {
	struct rrr_mqtt_subscription_collection topics;
	char *broker;
	uint16_t port;

	uint8_t nice_json;

	uint8_t use_v3;
	uint8_t use_v5;

	uint8_t ban_v5;

	uint8_t qos;

	struct rrr_mqtt_property_collection connect_properties;
	struct rrr_mqtt_client_data *mqtt_client;

	int transport_handle;
	struct rrr_mqtt_session *session;

	uint64_t disconnect_time;
	uint64_t subscriptions_send_time;
};

static void __rrr_sub_data_cleanup (
		struct rrr_sub_data *data
) {
	RRR_FREE_IF_NOT_NULL(data->broker);
	rrr_mqtt_client_destroy(data->mqtt_client);
	rrr_mqtt_subscription_collection_clear(&data->topics);
	rrr_mqtt_property_collection_clear(&data->connect_properties);
}

static int __rrr_sub_suback_unsuback_handler (struct rrr_mqtt_client_data *client, struct rrr_mqtt_p_suback_unsuback *packet, void *arg) {
	struct rrr_sub_data *data = arg;

	(void)(client);
	(void)(packet);
	(void)(data);

	return RRR_MQTT_OK;
}

static int __rrr_sub_packet_parsed_handler (struct rrr_mqtt_client_data *client, struct rrr_mqtt_p *packet, void *arg) {
	struct rrr_sub_data *data = arg;

	(void)(client);
	(void)(packet);
	(void)(data);

	return RRR_MQTT_OK;
}

static void __rrr_sub_receive_publish (struct rrr_mqtt_p_publish *publish, void *arg) {
	struct rrr_sub_data *data = arg;

	struct rrr_msg *msg_tmp = NULL;
	rrr_length msg_target_size = 0;

	RRR_DBG_2("> Received PUBLISH with topic %s\n", publish->topic);

	if (rrr_msg_get_target_size_and_check_checksum (
			&msg_target_size,
			(const struct rrr_msg *) publish->payload->payload_start,
			publish->payload->size
	) == 0) {
		if (msg_target_size != publish->payload->size) {
			RRR_DBG_2("> Incorrect size or incomplete RRR message, ignoring\n");
			goto out;
		}

		if ((msg_tmp = rrr_allocate(publish->payload->size)) == NULL) {
			RRR_MSG_0("Warning: Failed to allocate %" PRIrrrl " bytes in %s\n", publish->payload->size, __func__);
			goto out;
		}
		memcpy(msg_tmp, publish->payload->payload_start, publish->payload->size);

		if (rrr_msg_dump_to_host_and_dump(msg_tmp, publish->payload->size) != 0) {
			RRR_MSG_0("Failed to dump RRR message\n");
		}
	}
	else if (publish->payload->size > 0) {
		if (data->nice_json) {
			struct rrr_json_object *json;
			char *data_tmp;
			if (rrr_json_object_parse_nolog(&json, publish->payload->payload_start, publish->payload->size) != 0) {
				goto not_json;
			}
			rrr_json_from_object_nolog_pretty(&data_tmp, json);
			rrr_log_printn_plain(data_tmp, strlen(data_tmp));
			rrr_free(data_tmp);
			goto out;
		}

		not_json:
		rrr_log_printn_plain(publish->payload->payload_start, publish->payload->size);
	}

	out:
	if (!rrr_config_global.do_json_output) {
		rrr_log_printn_plain("\n", 1);
	}
	RRR_FREE_IF_NOT_NULL(msg_tmp);
}

static int __rrr_sub_init (
		struct rrr_sub_data *data,
		struct cmd_data *cmd,
		struct rrr_event_queue *events,
		struct rrr_mqtt_common_init_data *mqtt_init_data
) {
	int ret = 0;

	// Broker address
	if ((data->broker = strdup(cmd_exists(cmd, "broker", 0)
		? cmd_get_value(cmd, "broker", 0)
		: "localhost"
	)) == NULL) {
		RRR_MSG_0("Failed to allocate memory for broker in %s\n", __func__);
		ret = 1;
		goto out;
	}

	// Broker port
	if ((ret = rrr_arguments_parse_port (
			&data->port,
			cmd,
			"port",
			1883
	)) != 0) {
		goto out;
	}

	// QoS
	if (cmd_exists(cmd, "qos", 0)) {
		if (cmd_exists(cmd, "qos", 1)) {
			RRR_MSG_0("Multiple --qos arguments was specified\n");
			ret = 1;
			goto out;
		}

		const char *qos = cmd_get_value(cmd, "qos", 0);
		if (strcmp(qos, "2") == 0) {
			data->qos = 2;
		}
		else if (strcmp(qos, "1") == 0) {
			data->qos = 1;
		}
		else if (strcmp(qos, "0") == 0) {
			data->qos = 0;
		}
		else {
			RRR_MSG_0("Invalid value '%s' for argument --qos given\n", qos);
			ret = 1;
			goto out;
		}
	}

	// Nice JSON output
	const int nice_json = cmd_exists(cmd, "nice-json", 0);
	if (nice_json) {
		data->nice_json = nice_json && 1;
	}

	// V3, V5 or both
	const int v3 = cmd_exists(cmd, "protocol-v3", 0);
	const int v5 = cmd_exists(cmd, "protocol-v5", 0);

	if (v3 || v5) {
		data->use_v3 = v3 != 0;
		data->use_v5 = v5 != 0;
	}
	else {
		data->use_v3 = 1;
		data->use_v5 = 1;
	}

	// Subscribe topics
	for (unsigned long i = 0; i <= RRR_MQTT_SUB_TOPICS_MAX; i++) {
		if (i == RRR_MQTT_SUB_TOPICS_MAX) {
			RRR_MSG_0("Too many topics given (max is %i)\n", RRR_MQTT_SUB_TOPICS_MAX);
			ret = 1;
			goto out_cleanup_broker;
		}
		if (cmd_exists(cmd, "topic", i)) {
			const char *topic = cmd_get_value(cmd, "topic", i);
			if ((ret = rrr_mqtt_subscription_collection_push_unique_str (
					&data->topics,
					topic,
					0,
					0,
					0,
					data->qos
			)) != 0) {
				goto out;
			}
		}
		else {
			break;
		}
	}

	if ((ret = rrr_mqtt_client_new (
			&data->mqtt_client,
			mqtt_init_data,
			events,
			rrr_mqtt_session_collection_ram_new_client,
			NULL,
			__rrr_sub_suback_unsuback_handler,
			data,
			__rrr_sub_packet_parsed_handler,
			data,
			__rrr_sub_receive_publish,
			data
	)) != 0) {
		goto out_cleanup_topic;
	}

	goto out;
	out_cleanup_topic:
		rrr_mqtt_subscription_collection_clear(&data->topics);
	out_cleanup_broker:
		RRR_FREE_IF_NOT_NULL(data->broker);
	out:
		return ret;
}

static volatile int main_running = 1;
static volatile int sigusr2 = 0;

int rrr_signal_handler(int s, void *arg) {
	return rrr_signal_default_handler(&main_running, &sigusr2, s, arg);
}

static int __rrr_sub_periodic(RRR_EVENT_FUNCTION_PERIODIC_ARGS) {
	struct rrr_sub_data *data = arg;

	int alive = 0;
	int send_allowed = 0;
	int close_wait = 0;

	int ret_tmp;

	if (data->transport_handle > 0 && (ret_tmp = rrr_mqtt_client_connection_check_alive (
			&alive,
			&send_allowed,
			&close_wait,
			data->mqtt_client,
			data->transport_handle
	)) != 0) {
		printf("Alive: %i\n", ret_tmp);
		return RRR_EVENT_ERR;
	}

	if (data->disconnect_time > 0) {
		if (!alive && !close_wait) {
			RRR_DBG_1("| Disconnect complete\n");
			return RRR_EVENT_EXIT;
		}
		else if (rrr_time_get_64() > data->disconnect_time + RRR_MQTT_DISCONNECT_TIMEOUT_S * 1000 * 1000) {
			RRR_DBG_1("| Disconnect timeout\n");
			return RRR_EVENT_EXIT;
		}
	}
	else if (!alive && main_running) {
		RRR_DBG_1("| Connecting to broker...\n");

		data->subscriptions_send_time = 0;
		data->disconnect_time = 0;

		const uint8_t protocol_version = data->ban_v5 && data->use_v3
			? 4
			: data->use_v5
				? 5
				: 4
		;

		// Ban v5 every other connection attempt
		data->ban_v5 = !data->ban_v5;

		if ((ret_tmp = rrr_mqtt_client_connect (
				&data->transport_handle,
				&data->session,
				data->mqtt_client,
				data->broker,
				(uint16_t) data->port,
				protocol_version,
				30,   // Keep-alive
				1,    // Clean start,
				NULL, // Username
				NULL, // Password
				&data->connect_properties,
				NULL, // Will topic
				NULL, // Will message
				0,    // Will QoS
				0     // Will retain
		)) != 0) {
			if (ret_tmp == RRR_MQTT_SOFT_ERROR) {
				RRR_MSG_0("Connection to %s:%i failed, trying again\n",
					data->broker, data->port);
			}
			else {
				RRR_MSG_0("Connection to %s:%i failed, critical error\n",
					data->broker, data->port);
				return RRR_EVENT_ERR;
			}
		}
	}
	else if (send_allowed) {
		if (!main_running) {
			RRR_DBG_1("| Disconnecting...\n");
			data->disconnect_time = rrr_time_get_64();

			if ((ret_tmp = rrr_mqtt_client_disconnect (
					data->mqtt_client,
					data->transport_handle,
					0 // Reason
			)) != 0 && ret_tmp != RRR_MQTT_INCOMPLETE) {
				return RRR_EVENT_ERR;
			}
		}
		else if (!data->subscriptions_send_time) {
			RRR_DBG_1("| Sending subscriptions...\n");
			data->subscriptions_send_time = rrr_time_get_64();

			if (rrr_mqtt_client_subscribe (
					data->mqtt_client,
					&data->session,
					&data->topics
			) != 0) {
				return RRR_EVENT_ERR;
			}
		}
	}
	else if (!main_running) {
		RRR_DBG_1("| Exiting\n");
		return RRR_EVENT_EXIT;
	}

	if (sigusr2) {
		RRR_MSG_0("Received SIGUSR2, but this is not implemented in subscriber\n");
		sigusr2 = 0;
	}

	return RRR_EVENT_OK;
}

int main (int argc, const char **argv, const char **env) {
	if (!rrr_verify_library_build_timestamp(RRR_BUILD_TIMESTAMP)) {
		fprintf(stderr, "Library build version mismatch.\n");
		exit(EXIT_FAILURE);
	}

	int ret = EXIT_SUCCESS;

	struct rrr_signal_handler *signal_handler = NULL;
	struct rrr_net_transport_config net_transport_config = RRR_NET_TRANSPORT_CONFIG_PLAIN_INITIALIZER;

	if (rrr_allocator_init() != 0) {
		ret = EXIT_FAILURE;
		goto out_destroy_net_transport_config;
	}

	if (rrr_log_init() != 0) {
		ret = EXIT_FAILURE;
		goto out_cleanup_allocator;
	}

	rrr_strerror_init();

	rrr_signal_default_signal_actions_register();
	signal_handler = rrr_signal_handler_push(rrr_signal_handler, NULL);
	rrr_signal_handler_set_active (RRR_SIGNALS_ACTIVE);

	struct cmd_data cmd;
	struct rrr_sub_data data = {0};
	struct rrr_event_queue *events = NULL;
	struct rrr_mqtt_common_init_data mqtt_init_data = {
		NULL, /* Client name */
		RRR_MQTT_COMMON_RETRY_INTERVAL_S * 1000 * 1000,
		RRR_MQTT_COMMON_CLOSE_WAIT_TIME_S * 1000 * 1000,
		RRR_MQTT_COMMON_MAX_CONNECTIONS
	};

	cmd_init(&cmd, cmd_rules, argc, argv);

	if (rrr_main_parse_cmd_arguments_and_env(&cmd, env, CMD_CONFIG_DEFAULTS) != 0) {
		ret = EXIT_FAILURE;
		goto out_cleanup_allocator;
	}

	if (rrr_main_print_banner_help_and_version(&cmd, 1) != 0) {
		goto out_cleanup_cmd;
	}

	if (rrr_event_queue_new(&events) != 0) {
		ret = EXIT_FAILURE;
		goto out_cleanup_cmd;
	}

	if (__rrr_sub_init (
			&data,
			&cmd,
			events,
			&mqtt_init_data
	) != 0) {
		ret = EXIT_FAILURE;
		goto out_destroy_event;
	}

	if (rrr_mqtt_client_start(data.mqtt_client, &net_transport_config) != 0) {
		ret = EXIT_FAILURE;
		goto out_cleanup_data;
	}

	if (rrr_event_dispatch (
			events,
			250 * 1000, // 250 ms
			__rrr_sub_periodic,
			&data
	) != 0) {
		ret = EXIT_FAILURE;
	}

	rrr_signal_handler_set_active(RRR_SIGNALS_NOT_ACTIVE);
	rrr_signal_handler_remove(signal_handler);
	rrr_config_set_debuglevel_on_exit();

	out_cleanup_data:
		__rrr_sub_data_cleanup(&data);
	out_destroy_event:
		rrr_event_queue_destroy(events);
	out_cleanup_cmd:
		cmd_destroy(&cmd);
		rrr_socket_close_all();
		rrr_strerror_cleanup();
	out_cleanup_allocator:
		rrr_allocator_cleanup();
	out_destroy_net_transport_config:
		rrr_net_transport_config_cleanup(&net_transport_config);
		return ret;
}
